feat: 유저 정보 포함한 글 관리

jamie9504 opened the pull request:
**이슈 번호**

resolved: #10 

**작업 내용**

1. 유저 정보 포함한 글 관리

**테스트 작성 여부**

- [X] Test case

**주의 사항**

jamie9504 committed:
feat: Post 인수 테스트에 로그인 과정 추가

jamie9504 committed:
refactor: 모든 자원에 대해, 접근시 USER 권한 필요하도록 변경

jamie9504 committed:
feat: acceptance 테스트시 user 테이블 truncate

jamie9504 committed:
feat: Post Service에 Authentication 파라미터 추가

jamie9504 committed:
feat: Post와 User 연관관계 매핑 및 Post 서비스에 권한 적용

- lombok 적용
- Post와 User 연관관계 매핑
- Post Service에 권한(Authorization) 적용

jamie9504 committed:
feat: NotAuthorizedException Handler 추가

- PostServiceTest Bad Case 추가
- PostControllerTest Bad Case 추가

jamie9504 committed:
refactor: 사용하지 않는 애너테이션 제거

jamie9504 committed:
refactor: 사용하지 않는 생성자 및 애너테이션 제거

ttkmw reviewed (approved):
확인했습니다 어프루브합니다~

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/GlobalExceptionHandler.java:
```diff
@@ -26,6 +27,13 @@
+    @ExceptionHandler(NotAuthorizedException.class)
+    public ResponseEntity<ErrorResponse> handleForbidden(NotAuthorizedException e) {
+        return ResponseEntity
+            .status(HttpStatus.FORBIDDEN)
```
NotAuthorized 라면 403보다는 401(Unauthorized)이 낫지 않을까요?

jamie9504 replied:
Authorization의 경우 권한을 의미하는데요!

아래 글을 읽어보시면 403이 더 적합하다는 것을 알 수 있습니다 :)

`401 Unauthorized(권한 없음): 인증이 필요한 리소스에 인증 없이 접근할 경우 발생한다. 이 응답 코드를 사용할 때에는 반드시 브라우저에 어느 인증 방식[2]을 사용할 것인지 보내야 한다. 단순히 권한이 없는 경우 이 응답 코드 대신 아래 403 Forbidden을 사용해야 한다.`

aegis1920 replied:
덕분에 401과 403에 대해서 더 정확히 알 수 있었습니다 👍 
아래 이해하기 쉬운 예시가 있어서 적어놓았습니다!
> 쉽게 정리하자면, 401은 익명의 사용자, 403는 로그인은 하였으나 권한이 없는 사용자 정도로 구분하면 될 것 같습니다.

출처 : https://kimdoky.github.io/http/2018/12/30/401_403/

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/BaseEntity.java:
```diff
@@ -6,12 +6,23 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
+@Getter
+@EqualsAndHashCode(of = "id")
```
id 식별자가 있는데 id필드만 따로 Equals와 HashCode를 오버라이드한 이유가 있을까요?

jamie9504 replied:
이 부분 찾아봤던 문제인데요,

id 식별자가 있는 경우 영속 상태에서는 해당 값들이 같은지 구분이 되지만
준영속인 상태에서는 구분이 되지 않아서 어떻게 해야할 지 확인해보았는데, 이럴 경우 식별자를 대상으로 Equals와 HashCode를 오버라이딩해주면 된다고 해서 적용했습니다!

필요했던 상황은 getPrincipal()해서 받았던 User와 DB에서 조회해서 가져왔던 User를 비교할 때, 였습니다.

aegis1920 replied:
준영속 상태일 때 다르군요!! 알려주셔서 감사합니다 👍 👍

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/BaseEntity.java:
```diff
@@ -6,12 +6,23 @@
 @MappedSuperclass
 @EntityListeners(AuditingEntityListener.class)
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor(access = AccessLevel.PROTECTED)
```
BaseEntity를 상속한 클래스는 객체를 생성할 때 항상 id, createdAt, modifiedAt을 null값으로 둬야 하는데 `@AllArgsConstructor`를 해준 이유가 있을까요?

jamie9504 replied:
authentication을 통해 User를 Mocking할 때, id를 임의로 지정해주는 부분이 있어서 해당 부분 생성자를 열어뒀습니다!

``` java
// PostServiceTest.java
when(authentication.getPrincipal()).thenReturn(user);
```

jamie9504 replied:
이 부분 히히랑 얘기해서 수정해보겠습니다!

aegis1920 replied:
테스트를 위해서 프로덕션 코드에 @AllArgsConstructor를 넣는 것은 안된다고 생각합니다! 또한 더 정확한 테스트를 위해 의존성이 증가하더라도 id를 임의로 지정해주는 부분을 없애고 User를 조회해서 테스트하는 방법이 더 옳다고 생각이 듭니다 :)

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/Post.java:
```diff
@@ -16,13 +28,15 @@
-    protected Post() {
-    }
+    @ManyToOne
+    @JoinColumn(name = "USER_ID")
```
예약어들과 구분짓기 위해 테이블, 뷰, 컬럼을 비롯한 모든 식별자들은 소문자로 작성하는 게 좋다고 하네요 :) ORM이 테이블을 만들어줄 때 소문자로 만드는 것처럼요!

jamie9504 replied:
그렇군요! 많은 DB에서 대/소문자를 구분하지 않기 때문에 신경을 쓰지 않았었는데 소문자로 변경해두겠습니다~!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/PostService.java:
```diff
@@ -47,14 +51,26 @@ private Post findNotDeletedPost(Long id) {
+    private void validateIsOwner(Authentication authentication, Post post) {
+        User user = (User) authentication.getPrincipal();
+
+        if (!post.getUser().equals(user)) {
```
`getUser()` 하지 않고 메시지를 보내도 되지 않을까요? 그리고 `!`도 메서드 명에 Not을 붙여서 메서드로 빼도 될 것 같아요 :)

jamie9504 replied:
메시지를 보내도록 메서드를 생성하였습니다. 

`public boolean isNotSame(User user) {`

aegis1920 commented on back-end/legeno-around-here/src/test/resources/truncate.sql:
```diff
@@ -0,0 +1,6 @@
+DELETE
+FROM POST;
+DELETE
+FROM USER_ROLES;
+DELETE
+FROM USER;
```
TRUNCATE가 아닌 DELETE를 사용한 이유가 있을까요?

jamie9504 replied:
TRUNCATE의 경우, FK가 있는 상황에서는 제약조건에 걸려 테이블을 삭제하지 못하는 부분이 있어서

DELETE를 사용하던지 제약조건을 풀어줬다가 다시 제약조건을 걸어두던지 해야하는데요,
그 부분에 있어서는 DELETE가 더 안전할 것이라고 생각했습니다~!

jamie9504 replied:
파일명을 `init-table.sql`로 변경했습니다!

aegis1920 replied:
FK가 있을 경우, `SET FOREIGN_KEY_CHECKS = 0;` 로 외래키 제약을 비활성화 해줄 수 있긴 합니다만 이로 인해 불완전한 데이터가 발생할 수 있기에 좋은 방법은 아닌 것 같습니다. DELETE로 사용할 거라면 파일명을 truncate.sql이 아니라 다른 파일명으로 바꿔야 할 것 같아요 :)

aegis1920 reviewed (commented):
Post와 User를 합치는데 고생 많으셨습니다! 굿굿 👍
피드백 드렸으니 확인해주세요 :)

jamie9504 committed:
refactor: 컬럼명 소문자 변환

jamie9504 committed:
refactor: getter를 사용하지 않고 객체에 메시지를 보내도록 수정

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/PostController.java:
```diff
@@ -26,8 +27,9 @@ public PostController(PostService postService) {
     @PostMapping
-    public ResponseEntity<Void> createPost(@RequestBody PostRequest postRequest) {
-        Long postId = postService.createPost(postRequest).getId();
+    public ResponseEntity<Void> createPost(Authentication authentication,
```
현재 create, update, delete 함수에서 파라미터로 Authentication를 받고 별도의 작업 없이 service 넘겨주고 있는데요,
Authentication은 Security Filter Chain에서 SecurityContextHolder에 저장됩니다. 
단순히 Authentication을 얻기 위함이라면 Service쪽 코드에서 `SecurityContextHolder.getContext().getAuthentication()` 로 가져오는 것이 맞는 것 같습니다.

jamie9504 replied:
Good 👍 넵~!

jamie9504 committed:
Merge branch 'develop' into feature/#10

Hamliet commented on back-end/legeno-around-here/src/test/resources/truncate.sql:
```diff
@@ -0,0 +1,6 @@
+DELETE
```
truncate.sql 파일명을 다른 것으로 바꾸면 좋을 것 같아요

jamie9504 replied:
`init-table.sql`로 바꿔줬어요!

jamie9504 committed:
refactor: Table 초기화하는 파일명 변경

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/PostServiceTest.java:
```diff
@@ -71,13 +120,16 @@ void findAllPost_SuccessToFind() {
+            .findPost(createdPostResponse.getId());
 
         assertThat(updatedPostResponse.getWriting()).isEqualTo(updatedPostWriting);
+        assertThat(updatedPostResponse.getUser()).isEqualTo(UserResponse.from(user));
```
해당 assert구문이(유저가 변경되지 않았다는 것) 포스트 수정 테스트와 연관관계가 있을까요?

jamie9504 replied:
현재 Post 안에 필드로 writing과 user가 있는데요!

포스트를 수정하더라도 writing만 변경되는 것을 조금 더 명시적으로 해주고 싶었습니다 :)
(로직상 혹시 user를 날려버리거나 변경하는 것을 방지하기 위해...?)

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/PostServiceTest.java:
```diff
@@ -2,47 +2,96 @@
+    private Authentication authenticationOfAnother;
+
+    @BeforeEach
+    void setUp() {
```
위에서 언급한 것 처럼 service에서 SecurityContextHolder로 가져온다면 
SecurityContextHolder.getContext().getAuthentication().getPrincipal() 을 stubbing 하셔야 될 것 같네요

jamie9504 replied:
stubbing 하지 않고 하는 방법을 찾아서 적용해보았습니다~!

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/PostServiceTest.java:
```diff
@@ -107,19 +174,33 @@ void deletePost_HasId_SuccessToDelete() {
+        postService.deletePost(authentication, createdPostResponse.getId());
 
-        assertThatThrownBy(() -> postService.deletePost(createdPostResponse.getId()))
+        assertThatThrownBy(
```
개인적으론 엔터를 안 치는 편이 가독성이 더 좋아보이는데 이 부분은 같이 의견 나눠보아요 ㅎㅎ

jamie9504 replied:
이것은 길어서.. IDE가 자동으로 내려줍니다!

Hamliet reviewed (commented):
고생하셨습니다. 제히미!
코멘트 확인 부탁드려요~~!

jamie9504 committed:
refactor: 테스트 코드에서만 사용하는 생성자 제거

jamie9504 committed:
refactor: SecurityContextHolder를 서비스에서 직접 사용하도록 변경

- SecurityContextHolder(스프링 시큐리티 사용시 어디서든 사용할 수 있음)을 직접 서비스에서 사용하도록 변경
- Service Test Mocking하던 부분 SecurityContextHolder에 Authentication 설정하는 방식으로 변경

jamie9504 committed:
Merge branch 'feature/#10' of https://github.com/jamie9504/2020-legeno-around-here into feature/#10

Hamliet reviewed (approved):
고생하셨습니다! 이만 approve 할게요 :) 
👍 👍 👍

aegis1920 commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Test
+    void updatePost_IfNotAuthor_Forbidden() throws Exception {
```
여기서 말하는 Author가 작성자라면 Author 또는 Owner 둘 중 하나로 맞추는 게 좋을 것 같아요!

jamie9504 replied:
Product에 맞춰서, Owner로 맞춰놨어요!

jamie9504 committed:
test: product code와 메서드명 통일

aegis1920 reviewed (approved):
고생하셨습니다 천재이미~ 👍
